apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/profile/ProfileTaskChannelService.java

```diff
@@ -0,0 +1,243 @@
+                        if (getTaskListFuture != null) {
+                            getTaskListFuture.cancel(true);
+                        }
+
+                        // stop snapshot sender
+                        if (sendSnapshotFuture != null) {
+                            sendSnapshotFuture.cancel(true);
+                        }
+                    }
+                }
+            }
+        }
+
+    }
+
+    @Override
+    public void prepare() throws Throwable {
+        ServiceManager.INSTANCE.findService(GRPCChannelManager.class).addChannelListener(this);
+    }
+
+    @Override
+    public void boot() throws Throwable {
+        if (Config.Profile.ACTIVE) {
+            // query task list
+            getTaskListFuture = Executors.newSingleThreadScheduledExecutor(new DefaultNamedThreadFactory("ProfileGetTaskService"))
+                    .scheduleWithFixedDelay(new RunnableWithExceptionProtection(this, new RunnableWithExceptionProtection.CallbackWhenException() {
+                        @Override
+                        public void handle(Throwable t) {
+                            logger.error("Query profile task list failure.", t);
+                        }
+                    }), 0, Config.Collector.GET_PROFILE_TASK_INTERVAL, TimeUnit.SECONDS);
+
+            sendSnapshotFuture = Executors.newSingleThreadScheduledExecutor(new DefaultNamedThreadFactory("ProfileSendSnapshotService"))
+                    .scheduleWithFixedDelay(new RunnableWithExceptionProtection(new SnapshotSender(), new RunnableWithExceptionProtection.CallbackWhenException() {
+                        @Override public void handle(Throwable t) {
+                            logger.error("Profile segment snapshot upload failure.", t);
+                        }
+                    }), 0, 500, TimeUnit.MILLISECONDS);
+        }
+    }
+
+    @Override
+    public void onComplete() throws Throwable {
+    }
+
+    @Override
+    public void shutdown() throws Throwable {
+        if (getTaskListFuture != null) {
+            getTaskListFuture.cancel(true);
+        }
+
+        if (sendSnapshotFuture != null) {
+            sendSnapshotFuture.cancel(true);
+        }
+    }
+
+    @Override
+    public void statusChanged(GRPCChannelStatus status) {
+        if (GRPCChannelStatus.CONNECTED.equals(status)) {
+            Channel channel = ServiceManager.INSTANCE.findService(GRPCChannelManager.class).getChannel();
+            profileTaskBlockingStub = ProfileTaskGrpc.newBlockingStub(channel);
+            profileTaskStub = ProfileTaskGrpc.newStub(channel);
+        } else {
+            profileTaskBlockingStub = null;
+            profileTaskStub = null;
+        }
+        this.status = status;
+    }
+
+    /**
+     * add a new profiling snapshot, send to {@link #snapshotQueue}
+     */
+    public void addProfilingSnapshot(TracingThreadSnapshot snapshot) {
+        snapshotQueue.add(snapshot);
```

Comment:
if `Config.Profile.ACTIVE ==  false` TracingThreadSnapshot should be ignored, right?

Reply:
If not active profile, there will not going to happen.  I think it is useless to add judgment here, and it will also increase the cost of judgment.

Reply:
sorry, I was wrong.